### PR TITLE
Add Regression

### PR DESCRIPTION
**Pull request by nickwp**
This PR merges the regression branch with the main (classification) branch. Significant changes were needed to get everything working nicely together. Took this opportunity to also clean up the code that was being modified, but left other cleanup tasks for a future PR.

Summary of the merge:

- Features from classification branch now working for regression:
  - Double cover transform
  - Data augmentation
  - Choice of time and/or charge channels
  - Scaling of charge / time values
- Features from regression branchnow working for classification:
  - Choice of loss function and config for loss function parameters
  - Ability to run only on CPU instead of always requiring GPU
  - Ability to choose which quantity corresponds to the ground truth
- New features / fixes that came as a result of merging:
  - Much better logging to console and log files
  - Cleaned up config files
  - Removed old buggy models and other unused code
  - Fixes to update hydra version (suppresses hydra warnings)
  - Improved multi-GPU processing (faster and simpler using newer PyTorch procedures)
  - Fixed checkpointing bugs and correctly name the file by the choesn model and engine
  - Removed the need to specify the number of channels / classes in multiple places in the config
  - Config is simpler and more structured (but there's always more that could be done)
  - Simplified sampler code for working with multi-GPU
  - Updated default config to correspond to curent best IWCD training

Summary of code changes:

- Engine:
  - Introduced `RecontructionEngine` base class for the shared code, and `ClassificationEngine` and `RegressionEngine` derived classes for the code specific to classification and regression.
  - Added `truth_key` to allow choice of which key of the input corresponds to ground truth. For classification this will usually be "label" but for regression corresponds to whichever quantity we want to reconstruct. Ability to recontruct multiple different quantities together can be added in future.
  - Cleaned up and unified (across regression and classification) the logging to console and log files
  - Added ability to choose and configure loss function
  - Cleaned up and simplified DistributedDataParallel multi-GPU communication
  - General code cleanup
  - Fixed checkpointing file naming and saving
- Model:
  - Cleaned up old model files that are no longer used
  - Unified the two parts of classification network (the "feature extractor" and "classifier" so that only need to specify and set up a single model class to use. So the same ResNet / PointNet / etc can be used for either classification or regression with one config file that configures the chosen model.
  - Removed need to specify the number of channels / classes in multiple places in the config
- Dataset:
  - Changes to data augmentation to allow transforming truth values for regression
  - Refactor double cover transform to match changes to data augmentation transforms
  - Enable h5 dataset to load all the quantities needed for regression (in future should try to make it only load what is actually needed, to save a minute or two at the start of each training job)
  - Simplified sampler code for working with multi-GPU
- Analysis:
  - Update event display to work with dataset changes
  - Update classification and regression analysis to work with updated outputs
- Config:
  - Added default regression configs
  - Added default loss function configs
  - Simplification of model config to specify a single model, rather than separately defining sequential parts of the model
  - General clean up, fixed long-broken default configs, update to latest best hyperparameters

Tested for both classification (IWCD 4 class PID by @nickwp and IWCD multi-vertex vs single vertex by @adelorenzis) and regression (IWCD position, direction, energy by @nickwp). In all cases there are improvements (mostly small, but more significant for position regression)

### CODE REVIEW

**Comment by nickwp** [comment]
There appears to be a memory leak when running evaluation alone (after restore state). Doesn't seem to happen when running evaluation immediately after a training run. We need to investigate this.

File: analysis/classification.py [hunk 1]
```diff
@@ -432,10 +432,8 @@ def read_training_log_from_csv(self, directory):
             progression log
         """
         super().read_training_log_from_csv(directory)
-        self._train_log_accuracy = np.mean(self._log_train[:, :, 3], axis=0)
-        self._val_log_loss = self._log_val[:, 1]
```
Comment (felix-cormier) [inline analysis/classification.py:436]: Does the removal of loss here mean we lose the ability to plot? Or is this removal of duplicate code somewhere else?

File: watchmal/engine/reconstruction.py [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
+        """
+        # create the directory for saving the log and dump files
+        self.epoch = 0
+        self.step = 0
+        self.iteration = 0
+        self.best_validation_loss = 1.0e10
+        self.dump_path = dump_path
+        self.rank = rank
+        self.model = model
+        self.device = torch.device(gpu)
```
Comment (felix-cormier) [inline watchmal/engine/reconstruction.py:50]: Can this also be a CPU? As changed in the main function?

File: watchmal/engine/reconstruction.py [hunk 3]
```diff
@@ -0,0 +1,422 @@
...
+            print(f"  Validation {', '.join(f'{k}: {v:.5g}' for k, v in val_metrics.items())}", end="")
+            if val_metrics["loss"] < self.best_validation_loss:
+                print(" ... Best validation loss so far!")
+                self.best_validation_loss = val_metrics["loss"]
+                self.save_state(suffix="_BEST")
+                log_entries["saved_best"] = True
+            else:
+                print("")
+            # Save the latest model if checkpointing
+            if checkpointing:
```
Comment (felix-cormier) [inline watchmal/engine/reconstruction.py:307]: Is there only checkpointing every epoch now? Previous version would only save when validation loss was lower.

File: watchmal/engine/reconstruction.py
Comment (felix-cormier) [inline watchmal/engine/reconstruction.py:1]: A little hard to compare this one to the previous version since it's a newly named file, but I think overall everything looks consistent and with improvements

**Review by felix-cormier — APPROVED** [review]
Haven't had a chance to run code yet, but my review turned up no issues. Just a few comments & questions. I'll approve and hopefully test it later this week.

File: analysis/classification.py (hunk 1 above)
Comment (nickwp) [reply analysis/classification.py:436]: it's still available, just moved to the base class [here](https://github.com/WatChMaL/WatChMaL/pull/31/files/2ad1c240f5d184e947b361416e9a950a33a5d89d#diff-11dd39b5e5e4b11e71305b89d13495e88c2b55d96c11ddad010e093ff92c14fcR136) so it can be used for both classification and regression.

File: watchmal/engine/reconstruction.py (hunk 2 above)
Comment (nickwp) [reply watchmal/engine/reconstruction.py:50]: Yeah, thanks for spotting that. We could rename this parameter to `device` instead of `gpu` because it does actually work with CPU instead.

Looks like in main.py [here](https://github.com/WatChMaL/WatChMaL/pull/31/files/2ad1c240f5d184e947b361416e9a950a33a5d89d#diff-b10564ab7d2c520cdd0243874879fb0a782862c3c902ab535faabe57d5a505e1R109) it was renamed from `gpu` to `device` for that reason, so I can change it in the engine too for consistency.

Comment (nickwp) [reply watchmal/engine/reconstruction.py]: just pushed a commit that updated it from `gpu` to `device`

**Comment by Erwanlbv** [comment]
Tested on IWCD and HK for resnet and grant  (with a dump map between pmt ids and position for HK) and worked.
It worked. . There was only one minor warning that seems to be related to GitHub and hydra version

LGTM

File: config/reg_train.yaml
Comment (Erwanlbv) [inline config/reg_train.yaml:1]: Not tested regression. Only classification

File: config/reg_test.yaml
Comment (Erwanlbv) [inline config/reg_test.yaml:1]: Not tested regression. Only classification

File: watchmal/dataset/cnn_mpmt/cnn_mpmt_dataset.py
Comment (Erwanlbv) [inline watchmal/dataset/cnn_mpmt/cnn_mpmt_dataset.py:1]: I think implementing some much transforms in the mpmt makes it almost unreadable for new users. 
But it passed the test for iwcd and grant

File: watchmal/dataset/h5_dataset.py
Comment (Erwanlbv) [inline watchmal/dataset/h5_dataset.py:1]: We should probably create a folder for "base datasets" classed (like h5Dataset and RootDataset)

File: watchmal/engine/classification.py
Comment (Erwanlbv) [inline watchmal/engine/classification.py:1]: Does not support scalar target & model output

**Review by Erwanlbv — APPROVED** [review]
Tested on IWCD and HK for resnet and grant (with a dump map between pmt ids and positions for HK) and it worked.
There was only one minor warning that seems to be related to GitHub and hydra version.
This PR is good for me

**Comment by nickwp** [comment]
Thanks @felix-cormier and @Erwanlbv for the comments. Since this seems to be working I'll merge it now and we can make new issues / PRs for subsequent improvements, so those and be more easily added on top as needed.
